### PR TITLE
Move dry_run option higher in ordered option list

### PR DESCRIPTION
**Pull request by tubaxenor**
Currently `dry_run` option was set at the end of the options list, which makes `RSpec.configuration.dry_run?` always return `false` while processing required libs, especially `spec_helper.rb`.

For example in `spec_helper.rb`

```ruby
RSpec.configure do |config|
  puts "Dry?: #{config.dry_run?}"
end
```

```
$ rspec --dry-run

Dry? false
```

### CODE REVIEW

**Comment by pirj** [comment]
@tubaxenor May I kindly ask you to describe your use case, why do you need `dry_run?` to be available in the `configure` block?

File: lib/rspec/core/configuration_options.rb [hunk 1]
```diff
@@ -79,6 +79,10 @@ def order(keys)
         # deprecation (or otherwise access the reporter).
         :deprecation_stream,
 
+        # Dry run was passed through CLI args and must be set before `requires`
+        # to ensure RSpec.configuration.dry_run? is availble in all configure block.
```
Comment (JonRowe) [inline lib/rspec/core/configuration_options.rb:83]:
```suggestion
        # In order for `RSpec.configuration.dry_run?` to return `true` during
        # processing the `requires` option, it must be parsed before it.
```

**Review by JonRowe — CHANGES_REQUESTED** [review]
I'd reword the explanation slightly and can we get a spec check for this, but otherwise I'm happy with this change.

**Comment by tubaxenor** [comment]
> @tubaxenor May I kindly ask you to describe your use case, why do you need `dry_run?` to be available in the `configure` block?

Here is an abstract version of how we use this flag:

```ruby
RSpec.configure do |config|
  config.when_first_matching_example_defined(type: :system) do
    unless config.dry_run?
      require 'graphql/gateway_runner'
      Graphql::GatewayRunner.start
    end

    # Some other stuff runs at very first time a system spec was touched
  end
end
```

**Comment by tubaxenor** [comment]
> can we get a spec check for this, but otherwise I'm happy with this change.

Just pushed a commit to include `dry_run` option ordering 🙇

**Comment by pirj** [comment]
[`when_first_matching_example_defined` docs](https://relishapp.com/rspec/rspec-core/v/3-9/docs/hooks/when-first-matching-example-defined-hook) suggest performing expensive operations inside hooks:
```ruby
    RSpec.configure do |config|
      config.when_first_matching_example_defined(:db) do
        require "support/db"
      end
    end

And
    a file named "spec/support/db.rb" with:

    RSpec.configure do |config|
      config.before(:suite) do
        puts "Bootstrapped the DB."
      end

      config.around(:example, :db) do |example|
        puts "Starting a DB transaction."
        example.run
        puts "Rolling back a DB transaction."
      end
    end
```

While [`--dry-run` docs](https://relishapp.com/rspec/rspec-core/v/3-9/docs/command-line/dry-run-option) tell that hooks won't be run:

> Use the --dry-run option to have RSpec print your suite's formatter output **without running** any examples or **hooks**.

Would it work for you like this?

**Comment by tubaxenor** [comment]
I probably should provide more context about this, we are using kanpsack pro in queue mode to run our specs across CI nodes, which uses [dry run](https://github.com/KnapsackPro/knapsack_pro-ruby/blob/7b6398a9cf57cb46bf353acc0e5ac6c6a2074f4c/lib/knapsack_pro/test_case_detectors/rspec_test_example_detector.rb#L27) at the beginning of every process, which is why we do lots of `dry_run` checks to make sure we are not launching unnecessary services.

> Would it work for you like this?

It could work, but the whole point here is to not running extra services (in hooks or not) in dry run, shouldn't a `if config.dry_run?` makes more sense? Plus it's a CLI argument, it feels surprising (at least for me 😅 ) that `rspec --dry-run` returns a `false` value calling `RSpec.confoguration.dry_run?` in spec_helper or inside `RSpec.configure` block...

**Comment by pirj** [comment]
> the whole point here is to not running extra services (in hooks or not) in dry run

Right. That's what I suggest, putting your `Graphql::GatewayRunner.start` etc inside a `before(:suite)` that is conditionally defined depending on the presence of examples with a certain metadata in a portion of specs.

> shouldn't a `if config.dry_run?` makes more sense?

Seems explicit. Maybe overly. I admit I may have a bias towards the current state of things with hooks and metadata and against explicit.

> it feels surprising (at least for me 😅 ) that `rspec --dry-run` returns a `false` value when calling `RSpec.confoguration.dry_run?` in `spec_helper` or inside `RSpec.configure` block...

Indeed. Wondering if it's the only option like this.

In addition to this, there's `dry_run = ...` writer. Wouldn't it be surprising that it would override the setting from the command line?
Wondering if it does before or after your changes 🤔

**Comment by tubaxenor** [comment]
> In addition to this, there's dry_run = ... writer. Wouldn't it be surprising that it would override the setting from the command line?

Yes, it does feel wrong to have `dry_run` as a configurable option, but *maybe* someone wants that flexibility... 

> Wondering if it does before or after your changes 🤔
After I think? `process_options_into` is the first thing happen in `setup`.

**Comment by JonRowe** [comment]
Released in 3.13.0, apologies it has taken so long.
